wrapanapi/entities/server.py

```diff
@@ -0,0 +1,107 @@
+"""
+wrapanapi.entities.server
+
+Implements classes and methods related to actions performed on (physical) servers
+"""
+from abc import ABCMeta, abstractmethod
+
+from wrapanapi.entities.base import Entity
+
+
+class ServerState(object):
+    """
+    Represents a state for a server on the provider system.
+
+    Implementations of ``Server`` should map to these states
+    """
+    ON = 'ServerState.On'
+    OFF = 'ServerState.Off'
+    POWERING_ON = 'ServerState.PoweringOn'
+    POWERING_OFF = 'ServerState.PoweringOff'
+    UNKNOWN = 'ServerState.Unknown'
+
+    @classmethod
+    def valid_states(cls):
+        return [
+            var_val for _, var_val in vars(cls).items()
+            if isinstance(var_val, basestring) and var_val.startswith('ServerState.')
+        ]
+
+
+class Server(Entity):
+    """
+    Represents a single server on a management system.
+    """
+    __metaclass__ = ABCMeta
+    # Implementations must define a dict which maps API states returned by the
+    # system to a ServerState. Example:
+    #    {'On': ServerState.ON, 'Off': ServerState.OFF}
+    state_map = None
+
+    def __init__(self, *args, **kwargs):
+        """
+        Verify the required class variables are implemented during init
+
+        Since abc has no 'abstract class property' concept, this is the approach taken.
+        """
+        state_map = getattr(self, 'state_map')
+        if (not state_map or not isinstance(state_map, dict) or
+                not all(value in ServerState.valid_states() for value in state_map.values())):
+            raise NotImplementedError(
+                "property '{}' not properly implemented in class '{}'"
+                .format('state_map', self.__class__.__name__)
+            )
+        super(Server, self).__init__(*args, **kwargs)
+
+    def _api_state_to_serverstate(self, api_state):
+        """
+        Use the state_map for this instance to map a state string into a ServerState constant
+        """
+        try:
+            return self.state_map[api_state]
+        except KeyError:
+            self.logger.warn(
+                "Unmapped Server state '%s' received from system, mapped to '%s'",
+                api_state, ServerState.UNKNOWN
+            )
+            return ServerState.UNKNOWN
+
+    @abstractmethod
+    def _get_state(self):
```

Comment:
Can I suggest to replace `_get_state` with two functions:

```python
@property
def current_state(self)
  return self._api_state_to_serverstate(self.current_api_state)

@abstractproperty
def current_api_state(self):
  # to be implemented in subclass - it returns current server state from the API
```

I prefer `current_state` name because it better reflects what the function does - it returns current state. Also, we don't want every subclass to be calling the `_api_state_to_serverstate` if we can do it here already.

Reply:
We can't do that, because `_get_state` is prescribed by the ancestor class, so we have to implement it.

Reply:
[deleted]
looks like that ancestor class is what I define. I'll try something different then.